flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -222,12 +222,24 @@ protected StreamTask(
 
 	protected abstract void init() throws Exception;
 
-	protected abstract void run() throws Exception;
-
 	protected abstract void cleanup() throws Exception;
 
 	protected abstract void cancelTask() throws Exception;
 
+	/**
+	 * This method implements the default action of the task (e.g. processing one event from the input). Implementations
+	 * should (in general) be non-blocking.
+	 *
+	 * @return <code>true</code> if there is more work to perform as default action for this task and <code>false</code>
+	 * if the task is ready to finish.
+	 * @throws Exception on any problems in the action.
+	 */
+	protected abstract boolean performDefaultAction() throws Exception;
```

Comment:
I think you should either squash this commit with next one that introduces `ActionContext` and reverts/replaces those changes, or you should extract `ActionContext` and move it from 2nd to 1st commit.

Reply:
I would disagree with that because the context without the mailbox makes no sense and without the boolean return value and no mailbox this would not work (there may be ways to make it work, but those would then also have to be changed in a later commit, so no win). I think it is ok two have two commits that apply the change incrementally and can stand by themselves.

Reply:
Your changes are modifying the same lines of code twice, which just add reviewing noise (it took me some time to realise that you are overwriting those changes here later and I have already written a comment that something is missing with ` boolean performDefaultAction`)  and the same will be true in the future, if someone will want to look at the git history. 

Thus either we should come up with a sequence of changes that do not modify the same code twice too much (for example leaving `ActionContext` not implemented/empty/no-op in the first commit?) or the commits should be squashed.

Reply:
It changes the line twice because it is incrementally changed. I see no problem with that, it is work for different PRs with different purposes. If we would really decide to merge them, then we should also merge the JIRA tasks, but I think they are valuable as separated steps.

Even if i had a dummy context, the boolean is still required and the suggestion would not really help to solve the problem.

Reply:
In that case maybe let's squash them?

Reply:
Shall we close this PR and supersede it with https://github.com/apache/flink/pull/8431?